File: 0803-cheapest-flights-within-k-stops/0803-cheapest-flights-within-k-stops.cpp
class Solution {
public:
    int findCheapestPrice(int n, vector<vector<int>>& flights, int src, int dst,
                          int k) {
        vector<vector<pair<int, int>>> adj(n); // < to, cost > [] from index i

        for (auto it : flights) {
            adj[it[0]].push_back({it[1], it[2]});
        }

        queue<pair<int, pair<int, int>>> queue; // cost, <stops, node>
        vector<int> dist(n, 1e9);
        queue.push({0, {0, src}});
        dist[src] = 0;

        while (!queue.empty()) {
            auto top = queue.front();
            queue.pop();
            int cost = top.first;
            int stops = top.second.first;
            int node = top.second.second;

            if (stops > k)
                continue;

            for (auto it : adj[node]) {
                int adjNode = it.first;
                int edgeWt = it.second;

                if (cost + edgeWt < dist[adjNode] && stops <= k) {
                    dist[adjNode] = cost + edgeWt;
                    queue.push({cost + edgeWt, {stops + 1, adjNode}});
                }
            }
        }

        if (dist[dst] == 1e9)
            return -1;
        return dist[dst];
    }
};

File: 0803-cheapest-flights-within-k-stops/0803-cheapest-flights-within-k-stops.py
class Solution:
    def findCheapestPrice(self, n: int, flights: List[List[int]], src: int, dst: int, k: int) -> int:
        adj = defaultdict(list)

        cnt = 0
        for f, t, p in flights:
            if t == dst:
                cnt += 1
            adj[f].append((t, p))

        if cnt == 0:
            return -1

        queue = deque()
        price = [float("inf")]*n
        price[src] = 0

        queue.append((0, 0, src))

        while queue:
            cost, stops, node = queue.popleft()

            if stops > k:
                continue

            for adjNode, edgePrice in adj[node]:
                if price[adjNode] > cost + edgePrice:
                    price[adjNode] = cost + edgePrice
                    queue.append((cost+edgePrice, stops+1, adjNode))
        
        if price[dst] != float("inf"):
            return price[dst]
        return -1       

